net_util/src/tap.rs

```diff
@@ -559,127 +426,215 @@ mod tests {
-        let next_ip = NEXT_IP.fetch_add(1, Ordering::SeqCst);
-
-        let mut tap = Tap::new().unwrap();
-        let ip_addr = format!("{}{}", TAP_IP_PREFIX, next_ip).parse().unwrap();
+    fn construct_arp_reply<'a>(
+        buf: &'a mut [u8],
+        arp_frame: &EthernetFrame<&[u8]>,
+    ) -> EthernetFrame<'a, &'a mut [u8]> {
+        let arp_bytes = arp_frame.payload();
+        let arp_request = EthIPv4ArpFrame::request_from_bytes(arp_bytes).unwrap();
+        let mac_addr = MacAddr::parse_str(FAKE_MAC).unwrap();
+        let mut reply_frame =
+            EthernetFrame::write_incomplete(buf, arp_frame.src_mac(), mac_addr, ETHERTYPE_ARP)
+                .unwrap();
+        let sha = arp_request.sha();
+        let spa = arp_request.spa();
+        let tpa = arp_request.tpa();
+
+        EthIPv4ArpFrame::write_reply(
+            reply_frame.inner_mut().payload_mut(),
+            mac_addr,
+            tpa,
+            sha,
+            spa,
+        )
+        .unwrap();
+        reply_frame.with_payload_len_unchecked(ETH_IPV4_FRAME_LEN)
+    }
 
-        tap.set_ip_addr(ip_addr).unwrap();
+    fn make_tap(tap_ip: Ipv4Addr) -> Tap {
+        let tap = Tap::new().unwrap();
+        tap.set_ip_addr(tap_ip).unwrap();
         tap.set_netmask(SUBNET_MASK.parse().unwrap()).unwrap();
         tap.enable().unwrap();
+        tap
+    }
 
-        // Send a packet to the interface. We expect to be able to receive it on the associated fd.
-        pnet_send_packet(tap_name_to_string(&tap));
-
-        let mut buf = [0u8; 4096];
+    #[test]
+    fn test_read() {
+        // `fetch_add` adds to the current value, returning the previous value.
+        // reserve 2 fake IPs, one for the tap and the other for the virtual host
+        let next_ip = NEXT_IP.fetch_add(2, Ordering::SeqCst);
+        let tap_ip: Ipv4Addr = format!("{}{}", TAP_IP_PREFIX, next_ip).parse().unwrap();
+        let mut tap = make_tap(tap_ip);
+
+        // Now we want to set the target address to something that's near the IP address
+        // of the TAP (within its subnet) so the OS will think that the TAP is the next hop
+        // and forward the Udp packet through the TAP, where we can read it.
+        let dst_ip = format!("{}{}", TAP_IP_PREFIX, next_ip + 1).parse().unwrap();
+        let dst_port = 44445;
+        let dst_addr = SocketAddrV4::new(dst_ip, dst_port);
+
+        let src_port = 44444;
+        let src_addr = SocketAddrV4::new(tap_ip, src_port);
+        let socket = UdpSocket::bind(src_addr).expect("Failed to bind UDP socket");
+        socket.set_write_timeout(Some(Duration::new(5, 0))).unwrap();
+        socket.send_to(DATA_STRING.as_bytes(), dst_addr).unwrap();
 
         let mut found_packet_sz = None;
 
-        // In theory, this could actually loop forever if something keeps sending data through the
-        // tap interface, but it's highly unlikely.
+        // This will panic after a 5 sec timeout, or on receiving
```

Comment:
Just wondering, what causes this to panic after 5 seconds?

Reply:
oh, you're right. I should've used `set_read_timeout`, my bad. I'll fix that

Reply:
Oh, my bad. I edited this comment after finishing all the code and mixed it with the `test_write` functionality